python/dgl/transform.py

```diff
@@ -129,29 +132,37 @@ def knn_graph(x, k):
     (tensor([0, 1, 2, 2, 2, 3, 3, 3, 4, 5, 5, 5, 6, 6, 7, 7]),
      tensor([0, 1, 1, 2, 3, 0, 2, 3, 4, 5, 6, 7, 4, 6, 5, 7]))
     """
-    if F.ndim(x) == 2:
-        x = F.unsqueeze(x, 0)
-    n_samples, n_points, _ = F.shape(x)
-
-    dist = pairwise_squared_distance(x)
-    k_indices = F.argtopk(dist, k, 2, descending=False)
-    dst = F.copy_to(k_indices, F.cpu())
-
-    src = F.zeros_like(dst) + F.reshape(F.arange(0, n_points), (1, -1, 1))
-
-    per_sample_offset = F.reshape(F.arange(0, n_samples) * n_points, (-1, 1, 1))
-    dst += per_sample_offset
-    src += per_sample_offset
-    dst = F.reshape(dst, (-1,))
-    src = F.reshape(src, (-1,))
-    adj = sparse.csr_matrix(
-        (F.asnumpy(F.zeros_like(dst) + 1), (F.asnumpy(dst), F.asnumpy(src))),
-        shape=(n_samples * n_points, n_samples * n_points))
-
-    return convert.from_scipy(adj)
+    if use_dgl_impl:
+        if F.ndim(x) == 3:
+            x_size = tuple(F.shape(x))
+            x = F.reshape(x, (x_size[0] * x_size[1], x_size[2]))
+            x_seg = x_size[0] * [x_size[1]]
+        else:
+            x_seg = [F.shape(x)[0]]
+        return segmented_knn_graph(x, k, x_seg, use_dgl_impl=True)
+    else:
+        if F.ndim(x) == 2:
+            x = F.unsqueeze(x, 0)
+        n_samples, n_points, _ = F.shape(x)
+
+        dist = pairwise_squared_distance(x)
```

Comment:
Does nanoflann support bruteforcing?  If so this could be done in this PR.  If not and we still want to have two algorithms then I suggest to do it in a separate PR.

Reply:
nanoflann is a kd-tree implementation, it does not support sorting (brute force).

Reply:
Yep I agree to have it in another PR, and in that PR we should support both CPU and GPU.